File: caliber/CalibreEstEngine/backend/eva_service/src/agents/capabilities.py
"""Capability model — Python mirror of the shared JS capability map.

The authoritative definition lives in
    frontend/calibre-app/src/constants/capabilities.js
which the React app and the Node gateway both import directly. Python cannot
import JS, so this file mirrors it — and tests/test_capabilities_parity.py
parses the JS and asserts the two matrices are identical, so they cannot
silently diverge. That drift is not hypothetical: before this consolidation,
roles.js's PERMISSIONS granted rate-card access to all five roles while the
backend correctly denied it to estimator and senior_mgmt.

EVA uses this to decide which TOOLS a turn may reach. It is defence in depth,
not the only defence: upload-server re-checks the same capability on
/internal/estimator/scenario, so even a jailbroken model cannot execute a
scenario for a role that isn't permitted one.

RBAC model: 4 roles — admin, super, sme, estimator. `super` is the primary
Reviewer authority (holds ESTIMATE_APPROVE); there is no separate reviewer
role. A fifth role, senior_mgmt, existed historically and was removed —
role_can() fails closed for any role not in ROLE_CAPABILITIES, so an unknown
role (including a removed one) is never granted anything.
"""


class Capability:
    ESTIMATE_RUN = "estimate.run"
    ESTIMATE_VIEW_DETAIL = "estimate.view.detail"
    SCENARIO_RUN = "scenario.run"
    ESTIMATE_SAVE = "estimate.save"
    ESTIMATE_APPROVE = "estimate.approve"
    REVIEWER_ASSIGN = "reviewer.assign"
    PROJECT_CREATE = "project.create"
    RATE_CARD_VIEW = "ratecard.view"
    KNOWLEDGE_VIEW = "knowledge.view"
    KNOWLEDGE_UPLOAD = "knowledge.upload"
    KNOWLEDGE_REVIEW = "knowledge.review"
    EXECUTIVE_ANALYTICS_VIEW = "analytics.executive"
    PLATFORM_ANALYTICS_VIEW = "analytics.platform"
    TECHNICAL_REVIEW = "review.technical"
    USER_MANAGE = "users.manage"
    EVA_CHAT = "eva.chat"


ROLE_CAPABILITIES: dict[str, list[str]] = {
    "admin": [
        Capability.ESTIMATE_RUN, Capability.ESTIMATE_VIEW_DETAIL, Capability.SCENARIO_RUN,
        Capability.ESTIMATE_SAVE, Capability.ESTIMATE_APPROVE, Capability.REVIEWER_ASSIGN,
        Capability.PROJECT_CREATE,
        Capability.RATE_CARD_VIEW,
        Capability.KNOWLEDGE_VIEW, Capability.KNOWLEDGE_UPLOAD, Capability.KNOWLEDGE_REVIEW,
        Capability.EXECUTIVE_ANALYTICS_VIEW, Capability.PLATFORM_ANALYTICS_VIEW,
        Capability.TECHNICAL_REVIEW,
        Capability.USER_MANAGE,
        Capability.EVA_CHAT,
    ],
    "super": [
        Capability.ESTIMATE_RUN, Capability.ESTIMATE_VIEW_DETAIL, Capability.SCENARIO_RUN,
        Capability.ESTIMATE_SAVE, Capability.ESTIMATE_APPROVE, Capability.PROJECT_CREATE,
        Capability.RATE_CARD_VIEW,
        Capability.KNOWLEDGE_VIEW, Capability.KNOWLEDGE_UPLOAD,
        Capability.EXECUTIVE_ANALYTICS_VIEW,
        Capability.EVA_CHAT,
    ],
    "sme": [
        Capability.ESTIMATE_RUN, Capability.ESTIMATE_VIEW_DETAIL, Capability.SCENARIO_RUN,
        Capability.ESTIMATE_SAVE,
        Capability.RATE_CARD_VIEW,
        Capability.KNOWLEDGE_VIEW, Capability.KNOWLEDGE_UPLOAD, Capability.KNOWLEDGE_REVIEW,
        Capability.TECHNICAL_REVIEW,
        Capability.EVA_CHAT,
    ],
    "estimator": [
        Capability.ESTIMATE_RUN, Capability.ESTIMATE_VIEW_DETAIL, Capability.SCENARIO_RUN,
        Capability.ESTIMATE_SAVE,
        Capability.KNOWLEDGE_VIEW, Capability.KNOWLEDGE_UPLOAD,
        Capability.EVA_CHAT,
    ],
}


def role_can(role_id: str | None, capability: str) -> bool:
    """Unknown role or capability → False (fail closed)."""
    return capability in ROLE_CAPABILITIES.get(role_id or "", [])


# What EVA says instead of a bare refusal — each names what the role CAN still
# do, so a restricted turn stays useful (spec §24).
CAPABILITY_DENIAL_MESSAGE = {
    Capability.SCENARIO_RUN: (
        "Scenario execution is not available for this role. Explain that plainly, then offer what IS "
        "available: estimate health, cost drivers, risk drivers, and historical benchmarking. "
        "Do NOT estimate what a scenario would have produced."
    ),
    Capability.RATE_CARD_VIEW: (
        "Rate-card figures are restricted for this role. Do not state, paraphrase, derive, or "
        "approximate any day rate, blended rate, or margin — including by arithmetic from other "
        "numbers. Offer effort, FTE, risk and cost-driver analysis instead."
    ),
    Capability.ESTIMATE_RUN: (
        "Running a new estimate is not available for this role. Offer analysis of an existing "
        "estimate's outcome, risk and benchmarks instead."
    ),
    Capability.ESTIMATE_VIEW_DETAIL: (
        "The detailed component-level estimator view is not available for this role. Summarise at "
        "the executive level: effort, cost, FTE, risk band and planning range."
    ),
    Capability.ESTIMATE_SAVE: (
        "Creating, updating, or saving persisted estimates is not available for this role. Offer to "
        "discuss an existing saved estimate instead."
    ),
    Capability.ESTIMATE_APPROVE: (
        "Approving or rejecting a submitted estimate is not available for this role."
    ),
    Capability.REVIEWER_ASSIGN: (
        "Assigning a reviewer to an estimate is not available for this role."
    ),
    Capability.PROJECT_CREATE: (
        "Creating a project from an approved estimate is not available for this role."
    ),
}


def denial_message(capability: str) -> str:
    return CAPABILITY_DENIAL_MESSAGE.get(
        capability, "That action is not available for this role."
    )

File: caliber/CalibreEstEngine/frontend/calibre-app/src/constants/capabilities.js
// ─── Capability model — the single source of truth for "who may do what" ────
//
// This file replaces three parallel, mutually-contradicting role models that
// existed before: roles.js's PERMISSIONS matrix (46 keys, entirely dead code),
// roles.js's CARD_ACCESS, and workflows.js's allowedRoles/requiresAdmin. Their
// disagreement was not academic — PERMISSIONS granted 'eva.ratecards' to all
// roles while the backend correctly denied rate cards to estimator (and, at
// the time, the now-removed senior_mgmt role).
//
// RBAC model: 4 roles — admin, super, sme, estimator. `super` is the primary
// Reviewer authority (holds ESTIMATE_APPROVE); there is no separate reviewer
// role. A fifth role, senior_mgmt, existed historically and was removed —
// unknown roles fail closed (roleCan returns false), never granted access.
//
// IMPORTANT: this module is imported by BOTH the React app and the Node
// gateway (backend/upload-server), the same way scenarioRunner.js imports the
// estimator engine — so the browser and the server cannot drift apart. It is
// mirrored for Python in eva_service/src/agents/capabilities.py, and a pytest
// asserts the two stay identical.
//
// The frontend uses this for UX (what to show). The SERVER is always the
// authority — every sensitive route re-checks with requireCapability(), so a
// tampered browser gains nothing.

export const CAPABILITIES = {
  // Estimation
  ESTIMATE_RUN: 'estimate.run',                    // execute Layer 1 + /score
  ESTIMATE_VIEW_DETAIL: 'estimate.view.detail',    // 67-component workspace
  SCENARIO_RUN: 'scenario.run',                    // what-if execution
  ESTIMATE_SAVE: 'estimate.save',                  // create/update/save persisted estimates (own)
  ESTIMATE_APPROVE: 'estimate.approve',            // review decisions (approve/reject/request changes)
  REVIEWER_ASSIGN: 'reviewer.assign',              // assign a reviewer to a submitted estimate
  // Commercially sensitive
  RATE_CARD_VIEW: 'ratecard.view',                 // day rates / blended rates
  // Knowledge Hub
  KNOWLEDGE_VIEW: 'knowledge.view',
  KNOWLEDGE_UPLOAD: 'knowledge.upload',
  KNOWLEDGE_REVIEW: 'knowledge.review',            // edit/curate/see raw data docs
  // Analytics
  EXECUTIVE_ANALYTICS_VIEW: 'analytics.executive',
  PLATFORM_ANALYTICS_VIEW: 'analytics.platform',   // model + system health
  TECHNICAL_REVIEW: 'review.technical',            // SME validation view
  // Governance
  USER_MANAGE: 'users.manage',
  // Assistant
  EVA_CHAT: 'eva.chat',
};

const C = CAPABILITIES;

// Seeded to EXACTLY the behaviour the server already enforced before this
// refactor, so consolidation changes no access decision:
//   /api/score + /internal/estimator/scenario → admin, super, sme, estimator
//   rate cards (RATE_CARD_RESTRICTED_ROLES)   → denied to estimator
//   /api/auth/users*                          → admin only
export const ROLE_CAPABILITIES = {
  admin: [
    C.ESTIMATE_RUN, C.ESTIMATE_VIEW_DETAIL, C.SCENARIO_RUN,
    C.ESTIMATE_SAVE, C.ESTIMATE_APPROVE, C.REVIEWER_ASSIGN,
    C.RATE_CARD_VIEW,
    C.KNOWLEDGE_VIEW, C.KNOWLEDGE_UPLOAD, C.KNOWLEDGE_REVIEW,
    C.EXECUTIVE_ANALYTICS_VIEW, C.PLATFORM_ANALYTICS_VIEW, C.TECHNICAL_REVIEW,
    C.USER_MANAGE,
    C.EVA_CHAT,
  ],
  super: [
    C.ESTIMATE_RUN, C.ESTIMATE_VIEW_DETAIL, C.SCENARIO_RUN,
    C.ESTIMATE_SAVE, C.ESTIMATE_APPROVE,
    C.RATE_CARD_VIEW,
    C.KNOWLEDGE_VIEW, C.KNOWLEDGE_UPLOAD,
    C.EXECUTIVE_ANALYTICS_VIEW,
    C.EVA_CHAT,
  ],
  sme: [
    C.ESTIMATE_RUN, C.ESTIMATE_VIEW_DETAIL, C.SCENARIO_RUN,
    C.ESTIMATE_SAVE,
    C.RATE_CARD_VIEW,
    C.KNOWLEDGE_VIEW, C.KNOWLEDGE_UPLOAD, C.KNOWLEDGE_REVIEW,
    C.TECHNICAL_REVIEW,
    C.EVA_CHAT,
  ],
  estimator: [
    C.ESTIMATE_RUN, C.ESTIMATE_VIEW_DETAIL, C.SCENARIO_RUN,
    C.ESTIMATE_SAVE,
    C.KNOWLEDGE_VIEW, C.KNOWLEDGE_UPLOAD,
    C.EVA_CHAT,
  ],
};

/** Does this role hold this capability? Unknown role/capability → false. */
export function roleCan(roleId, capability) {
  return ROLE_CAPABILITIES[roleId]?.includes(capability) ?? false;
}

/** Every capability a role holds (used by the nav/dashboard config). */
export function capabilitiesFor(roleId) {
  return ROLE_CAPABILITIES[roleId] ?? [];
}

// Human-readable denial copy. A bare "Access Denied" is unhelpful; each of
// these names what the role CAN still do, so the UI and EVA can degrade
// gracefully instead of dead-ending.
export const CAPABILITY_DENIAL_MESSAGE = {
  [C.ESTIMATE_RUN]: 'Running estimates isn\'t available for your role. You can still review estimate outcomes, risk and historical benchmarks.',
  [C.ESTIMATE_VIEW_DETAIL]: 'The detailed estimator workspace isn\'t available for your role. The executive summary shows effort, cost, FTE and risk.',
  [C.SCENARIO_RUN]: 'Scenario execution isn\'t available for your role. You can still view estimate health, cost drivers and risk analysis.',
  [C.ESTIMATE_SAVE]: 'Creating or saving estimates isn\'t available for your role. You can still view and discuss existing estimates.',
  [C.ESTIMATE_APPROVE]: 'Approving or rejecting submitted estimates is restricted to Admin and Delivery Leadership roles.',
  [C.REVIEWER_ASSIGN]: 'Assigning a reviewer to an estimate is restricted to platform administrators.',
  [C.RATE_CARD_VIEW]: 'Rate-card figures are restricted for your role. I can still explain effort, FTE, risk and the cost drivers behind an estimate.',
  [C.USER_MANAGE]: 'User management is restricted to platform administrators.',
  [C.KNOWLEDGE_REVIEW]: 'Curating Knowledge Hub content is restricted to Admin and SME roles.',
  [C.PLATFORM_ANALYTICS_VIEW]: 'Platform and model analytics are restricted to platform administrators.',
  [C.TECHNICAL_REVIEW]: 'The technical review workspace is available to SME and Admin roles.',
};

export function denialMessage(capability) {
  return CAPABILITY_DENIAL_MESSAGE[capability]
    || 'That action isn\'t available for your role.';
}
